django/db/models/fields/json.py

```diff
@@ -0,0 +1,111 @@
+import json
+
+from django import forms
+from django.core import exceptions
+from django.db.utils import NotSupportedError
+from django.utils.translation import gettext_lazy as _
+
+from . import Field
+from .mixins import CheckFieldDefaultMixin
+
+
+def convert_jsonfield_value(value, expression, connection):
+    if connection.vendor == 'oracle':
+        return value.read()
+    return value
+
+
+class JSONField(CheckFieldDefaultMixin, Field):
+    description = _('A JSON object')
+    default_error_messages = {
+        'invalid': _("Value must be valid JSON."),
+    }
+    _default_hint = ('dict', '{}')
+
+    def __init__(self, encoder=None, decoder=None, default=dict, *args, **kwargs):
+        self.encoder, self.decoder = encoder, decoder
+        super().__init__(default=default, *args, **kwargs)
+
+    def _check_json_support(self, connection):
+        if not connection.features.supports_json:
+            raise NotSupportedError(_('This database backend does not support JSONField.'))
```

Comment:
I would remove `_check_json_support()` and move check to `__init__()`.

Reply:
This would require me to import `connection` from `django.db`. If I use that `connection` to check for the `supports_json` feature, I get the following error on SQLite:

```
File "/tests/django/django/db/backends/sqlite3/base.py", line 173, in get_connection_params
    "settings.DATABASES is improperly configured. "
django.core.exceptions.ImproperlyConfigured: settings.DATABASES is improperly configured. Please supply the NAME value.
```

It's because the backend checks for the `NAME` in `get_connection_params`: https://github.com/django/django/blob/master/django/db/backends/sqlite3/base.py#L169

However, the included `test_sqlite.py` doesn't supply that value: https://github.com/django/django/blob/master/tests/test_sqlite.py

I tried supplying the value myself (I just needed to supply it for the `default` database), and the tests pass.

Should I modify the test settings? If so, what should be the `NAME`? Should I also supply the `NAME` for the `other` database?

Also, is that the setting used by [djangoci.com](//djangoci.com)? If not, then it should be configured there, too.

Reply:
Instantiating a `cursor` to check for `supports_json` would call `ensure_connection` (and in turn `get_connection_params`) in the process. Since the `test_sqlite.py` settings don't have the `NAME` value for the database, the tests errored with the above message.

I found out the database name on [djangoci.com](/djangoci.com) after looking at the build output, and adjusted the settings accordingly:

```python
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'file:memorydb_default?mode=memory&cache=shared',
    },
    'other': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'file:memorydb_other?mode=memory&cache=shared',
    }
}
```

The tests pass now, but they still fail for the spatialite database. I'm guessing djangoci uses its own settings for that, but I'm not really sure...

Reply:
I think we should remove check from `JSONField.__init__()` and add a system check to `Model.check()` similar to the [_check_constraints()](https://github.com/django/django/blob/53209f78302a639032afabf5326d28d4ddd9d03c/django/db/models/base.py#L1814-L1835).

Reply:
I missed this, sorry. Done.